### PR TITLE
Speed outliers removal

### PR DESCRIPTION
**Pull request by bcaneco**
Hiya!

So, this covers the implementation of an alternative approach to remove outliers due to GPS glitches from the input dataset (and other sundries). This code amends a couple of problems with the the previous code, namely:
- incorrect removal of the location prior to the actual outlier, as it didn't account for the fact that mt_speed calculates speed to the next location (thus, the offending location is immediately after the location with abnormal speed)
- because it applied a single blanket filtering step (i.e. all locations with abnormal speeds filtered out at once), abnormal speeds would persist in the output under certain conditions.:
       - when more than 1 abnormal locations were recorded sequentially or
       - borderline threshold values

This code clean the data iteratively and chronologically, aiming at the correct detection and removal of glitched locations. I've written some [notes](https://github.com/dmpstats/Standardise_Formats_and_Calculate_Basic_Statistics/blob/speed-outliers-removal/dev/detect%20and%20remove%20outliers.md) demonstrating the new developed function at work.

Let me know what you think :)

### CODE REVIEW

**Comment by bcaneco** [comment]
Just noticed the first half of the commits are related to changes associated with the review in the other pull request. This is because it's tracking all the changes from the master branch. So please only consider the commits bellow [added function to detect and remove outliers](https://github.com/dmpstats/Standardise_Formats_and_Calculate_Basic_Statistics/pull/6/commits/070db1d51643934029f443e0c72f6aafcbb2d390). Sorry if this is a bit confusing!

File: RFunction.R [hunk 1]
```diff
@@ -398,3 +399,66 @@ rFunction = function(data,
...
+
+
+# //////////////////////////////////////////////////////////////////////////////
+# using speed (i.e. a combination of time and location) as a proxy to identify
+# outliers in movement data (due to e.g. GPS glitches). All location events
+# associated with speeds above the user-defined threshold (given by some
+# sensible estimate of nonsensical speeds for the tagged species) asre removed
+# from the data.
+
+drop_outliers <- function(data, tresh_kmph){
```
Comment (lindesaysh) [inline RFunction.R]: for consistency you might want to change `tresh` to `thresh` (just the missing h)

**Review by lindesaysh — APPROVED** [review]
this function and associated docs are great.  thanks Bruno!

File: dev/detect and remove outliers.qmd [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+  track_id_column = "track") |> 
+  sf::st_set_crs(4326L)
+
+dt_mv2_base
+```
+
+
+## Application to two consecutive outliers
+
+Add two outliers to the 2nd and 3rd positions and ad speeds (km/h).
```
Comment (lindesaysh) [inline dev/detect and remove outliers.qmd]: missing a "d" (add)

File: dev/detect and remove outliers.qmd [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+#| fig-cap: Points are location events, dashed arrows illustrate movement paths and labels denote movement speeds (km/h)
+#| label: fig-p2
+
+ggplot(dt_mv2_clean) +
+  geom_sf(col = "red", size = 3) +
+  geom_sf_label(aes(label = round(kmph, 1)), nudge_x = 0.04, nudge_y = -0.04, size = 2.5) +
+  geom_curve(
+    aes(x = X, y = Y, xend = lead(X), yend = lead(Y)), 
+    data = as_tibble(st_coordinates(dt_mv2_clean)),
+    curvature = -0.3,
```
Comment (lindesaysh) [inline dev/detect and remove outliers.qmd:128]: not a comment but the curves lines are pretty :)

File: RFunction.R (hunk 1 above)
Comment (bcaneco) [reply RFunction.R]: yep: typo-alert! 

Fixed in [7a066dd](https://github.com/dmpstats/Standardise_Formats_and_Calculate_Basic_Statistics/pull/6/commits/7a066dd340682a34f7420f2e393a593b81f428ef)

File: dev/detect and remove outliers.qmd (hunk 2 above)
Comment (bcaneco) [reply dev/detect and remove outliers.qmd]: Typo-alert!

Fixed in https://github.com/dmpstats/Standardise_Formats_and_Calculate_Basic_Statistics/pull/6/commits/371ad170ca546aa4b7252085a3814994bb328a19

File: dev/detect and remove outliers.qmd (hunk 3 above)
Comment (andreakoelzsch) [reply dev/detect and remove outliers.qmd:128]: Hi Bruno and Lindesay, thans for all the great coding here :) One idea that you have either already have implemented or not: We have previously adressed outlier removal by speed by doing a while loop until there are no more offending speeds, that way also e.g. groups of outliers are taken care of. THanks! Andrea

Comment (bcaneco) [reply dev/detect and remove outliers.qmd:128]: Hi Andrea, thanks for having a look at this and for the feedback!

Yes, I think that's same/similar approach we take in our speed-outlier removal function :)

https://github.com/dmpstats/Standardise_Formats_and_Calculate_Basic_Statistics/blob/371ad170ca546aa4b7252085a3814994bb328a19/RFunction.R#L406-L463

**Review by callumjclarke — APPROVED** [review]

**Pull request merged**